### PR TITLE
Upgrade unipi to paf-le-chien

### PR DESCRIPTION
**Pull request by dinosaure**
It's a draft, at least it compiles but the TLS path seems buggy and it's hard to understand why. I will try to investigate more deeper on this side soon. But it's a proposal, feel free to share your opinion on it and improve this unikernel :+1:.

### CODE REVIEW

**Comment by dinosaure** [comment]
So an implementation of `unipi` with `paf` is running here: https://unipi.egar.im/ (with a _debug_ let's encrypt certificate and small tweaks on `paf` available here: https://github.com/dinosaure/paf-le-chien/pull/28). WDYT?

**Comment by hannesm** [comment]
I don't know. What is the goal? This PR adds quite some boilerplate to the codebase :/

Will this remove the cohttp and conduit dependency entirely? if not, can we get to this point (easily?) -- maybe if let's encrypt specifies their own module type for the HTTP client used (to avoid the cohttp dependency)?

**Comment by dinosaure** [comment]
> I don't know. What is the goal? This PR adds quite some boilerplate to the codebase :/

I think indeed we can do something better on this side. The main problem is the `mimic`'s ritual needed to let the sub-module let's encrypt to safely communicate with let's encrypt and do the challenge. Such part is provided by `paf` with `paf.le` (see this functor: https://github.com/dinosaure/paf-le-chien/blob/master/lib/lE.mli

But I'm not sure that the sub-module `Letsencrypt` and `paf.le` are equivalent. I need to check.

> Will this remove the cohttp and conduit dependency entirely?

The only remaining module is the `Cohttp.Client.S` signature - then, `conduit` is definitely removed.

> maybe if let's encrypt specifies their own module type for the HTTP client used (to avoid the cohttp dependency)?

It can be a nicer solution indeed! Again, we already talk about `mirage-http` and an ability to provide such interface without any dependencies (with `cohttp` or `http/af`). If you think that is the best way, I can dig on this way 👍 .

On the other side, `paf` (it's not currently the case but should be easy to do) can handle ALPN and dispatch correctly HTTP 1.1 requests and HTTP 2.0 requests which can be interesting for us.

**Comment by dinosaure** [comment]
PS: I can try to run some benchmark to between this version and the version with `cohttp` to may be highlight an improvement 👍

**Comment by dinosaure** [comment]
So I did a large stress-test between `cohttp` and `http/af` and it seems that `http/af` can handle ~ 15 000 requests per sec when `cohttp` handles only ~ 2500 requests per sec for the same file. The magnitude is:
- for `http/af`, we need 0.007 sec to respond to the client
- for `cohttp`, we need 0.027 sec to respond to the client

Such test is done over TLS for both. This is the plain text of the benchmark (`http/af`):
```sh
dinosaure@turbine:~$ hey -n 1000000 -c 24 https://unipi.egar.im/

Summary:
  Total:	73.8878 secs
  Slowest:	0.2575 secs
  Fastest:	0.0010 secs
  Average:	0.0018 secs
  Requests/sec:	13533.8195
  
  Total data:	22999632 bytes
  Size/request:	23 bytes

Response time histogram:
  0.001 [1]	|
  0.027 [999926]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  0.052 [47]	|
  0.078 [9]	|
  0.104 [0]	|
  0.129 [0]	|
  0.155 [0]	|
  0.181 [0]	|
  0.206 [0]	|
  0.232 [0]	|
  0.258 [1]	|


Latency distribution:
  10% in 0.0013 secs
  25% in 0.0014 secs
  50% in 0.0015 secs
  75% in 0.0017 secs
  90% in 0.0021 secs
  95% in 0.0024 secs
  99% in 0.0103 secs

Details (average, fastest, slowest):
  DNS+dialup:	0.0000 secs, 0.0010 secs, 0.2575 secs
  DNS-lookup:	0.0000 secs, 0.0000 secs, 0.0183 secs
  req write:	0.0000 secs, 0.0000 secs, 0.0021 secs
  resp wait:	0.0017 secs, 0.0009 secs, 0.2575 secs
  resp read:	0.0000 secs, 0.0000 secs, 0.0025 secs

Status code distribution:
  [200]	999984 responses
```

And `cohttp`:
```sh
dinosaure@turbine:~$ hey -n 1000000 -c 24 https://unipi.egar.im/

Summary:
  Total:	427.5950 secs
  Slowest:	0.2574 secs
  Fastest:	0.0011 secs
  Average:	0.0102 secs
  Requests/sec:	2338.6243
  
  Total data:	22999632 bytes
  Size/request:	23 bytes

Response time histogram:
  0.001 [1]	|
  0.027 [828281]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  0.052 [170993]	|■■■■■■■■
  0.078 [681]	|
  0.104 [4]	|
  0.129 [0]	|
  0.155 [0]	|
  0.180 [0]	|
  0.206 [0]	|
  0.232 [0]	|
  0.257 [24]	|


Latency distribution:
  10% in 0.0019 secs
  25% in 0.0024 secs
  50% in 0.0028 secs
  75% in 0.0040 secs
  90% in 0.0451 secs
  95% in 0.0465 secs
  99% in 0.0495 secs

Details (average, fastest, slowest):
  DNS+dialup:	0.0000 secs, 0.0011 secs, 0.2574 secs
  DNS-lookup:	0.0000 secs, 0.0000 secs, 0.0019 secs
  req write:	0.0000 secs, 0.0000 secs, 0.0011 secs
  resp wait:	0.0020 secs, 0.0010 secs, 0.0456 secs
  resp read:	0.0081 secs, 0.0000 secs, 0.0680 secs

Status code distribution:
  [200]	999984 responses
```

The tool `hey` should be available. The client is located to another geographic place than the server. It a small benchmark and it's not really reproductible but it shows nice results about `http/af`.

**Comment by hannesm** [comment]
Thanks @dinosaure, the benchmark looks convincing.

About let's encrypt: I don't know what the state of mirage-http is. I'd be fine to have a HTTP Client module type in letsencrypt directly.

**Comment by talex5** [comment]
> So I did a large stress-test between cohttp and http/af and it seems that http/af can handle ~ 15 000 requests per sec when cohttp handles only ~ 2500 requests per sec for the same file.

That's a surprisingly large difference. There are some http benchmarks at https://github.com/ocaml-multicore/retro-httpaf-bench and there we see httpaf being "only" about twice as fast as cohttp:

mc-dev

That graph is from wrk2 with 1000 open connections repeatedly requesting a single static page (which fits in one packet). All servers were configured to use a single core for this test.

**Comment by dinosaure** [comment]
> That's a surprisingly large difference. There are some http benchmarks at https://github.com/ocaml-multicore/retro-httpaf-bench and there we see httpaf being "only" about twice as fast as cohttp:

Yeah, I think it's not a fair benchmark for many reasons:
1) I don't have any control on the flow between the server and the client (two differents servers on internet)
2) TLS is used in my context, not sure about its impact on performances (and how `cohttp`/`conduit` and `http/af`/`paf` differs on that point in details)
3) `mirage-tcpip` is used here
4) and the unikernel is virtualized with KVM - but `http/af` and `cohttp` are deployed into the same context

For these points, it's why I said that the "benchmark" is not reproductible and show-up (with `unipi`) especially a "real" usage about what we want (a simple MirageOS website synchronized to a Git repository). May be the context fits better for `http/af` but the huge diff between `cohttp` and `http/af` gives me some reasons to switch to `http/af` at the end 🙂 .

**Comment by dinosaure** [comment]
So https://unipi.egar.im/ is alive for a long time so I believe that I'm ready to cut a release of `paf` (which facilitates the way to get a let's encrypt certificate) and this PR will be ready to merge then!

**Comment by hannesm** [comment]
Thanks again @dinosaure -- there's still the outstanding question of the dependency cone in respect to letsencrypt (which currently has a hard dependency on cohttp, and paf has a dependency on letsencrypt). I'd appreciate if we can conclude:
- have a HTTP_client.S in letsencrypt (that is a module type which is trivially implemented by cohttp -- thus there's no need for any consumer change if you use a cohttp client)
- have letsencrypt.httpaf implement that module type
- remove (a) letsencrypt dependency from paf (or does it serve a good purpose there?) (b) have a unipi-without-cohttp [and then we can follow with other unikernels, removing the whole cohttp & conduit dependencies]

WDYT?

**Comment by dinosaure** [comment]
Yes, let me sometimes to shape all of that, I still need to think to revive `mirage-http` or just delete the hard dependency on `letsencrypt` about `cohttp` :+1:.

**Comment by hannesm** [comment]
So, now that letsencrypt 0.3.0 is in opam-repository, we could proceed with (a) adapting paf and (b) removing cohttp from unipi. This would clean up the dependency cone drastically :) (and reduce the binary size).

**Comment by dinosaure** [comment]
I updated the PR with last changes on:
- `ocaml-git` (we use `git-mirage.3.5.0` now)
- `letsencrypt` (we use `letsencrypt.3.0.0`)
- and `paf` & `paf-le` (we use `paf.0.0.5` & `paf-le.0.0.5`)

However, we still need to `pin` `irmin` with the right version. It seems that `irmin` will break its API and updates about it does not reflect yet what is going on `irmin~master` (the API will change again). However, at least, `unipi` compiles (I believe).

**Comment by hannesm** [comment]
with this branch, @hb9cwp reported successful builds (on Unix and hvt) on OpenBSD. But the hvt unikernel does not serve web pages -- maybe related to the httpaf/mimic/paf semantics of Lwt.async that were discussed above? @dinosaure did you do functional tests with a hvt unikernel (that it actually serves web pages)?

**Comment by hb9cwp** [comment]
@hannesm Good news: I have got both `unix` as well as `hvt` targets working now with this branch on both OpenBSD 6.9 and 7.0 after `opam update & upgrade` to latest, using `mirage v3.10.6` and `ocaml 4.10.2` :-)
Further, on OpenBSD 7.0, I had to raise the stack size from 4 to 32 kB using `ulimit -s 32768`, otherwise the make depend/build fail.
Finally, in my start script for the hvt unikernel, I corrected the netmask from /32 to /24 on the unikernel's `tap` interface which is bridged at layer-2 to the physical NIC of the OpenBSD host.
Tomorrow, I will try to apply the lessons learned to `dns-primary-git`.
Thank you for all the spontaneous support that got me so far!

**Comment by hannesm** [comment]
@hb9cwp great! This means your unix and hvt unikernels deliver data via HTTP(S) to clients that send requests (in contrast too your earlier comment that there's no content being delivered)?

**Comment by hb9cwp** [comment]
@hannesm Yes, exactly. So far, I tested serving simple .html pages and .png images with HTTP though, HTTPS clients to come. Also, Unipi's hook works and triggers it to refresh from Github using HTTPS, SSH to come.

File: unikernel.ml [hunk 1]
```diff
@@ -73,181 +83,153 @@ module Main (_ : sig end) (Http_client: Cohttp_lwt.S.Client) (Http: Cohttp_mirag
...
+    let hdr = Httpaf.Headers.add_unless_exists resp.Httpaf.Response.headers
+      "connection" "close" in
+    let resp = { resp with Httpaf.Response.headers= hdr } in
+    Httpaf.Reqd.respond_with_string reqd resp ""
+
   module Dispatch = struct
-    let dispatch store hookf hook_url request _body =
-      let p = Uri.path (Cohttp.Request.uri request) in
-      let path = if String.equal p "/" then "index.html" else p in
+    let dispatch store hookf hook_url _conn reqd =
```
[the remark this answers is not in the document]
Comment (dinosaure) [reply unikernel.ml:99]: `_conn` is the `Ipaddr.t * int` too. You can not remove it (because the caller pass it to the callback).

**Comment by hannesm** [comment]
@hb9cwp great, thanks for your confirmation.

> Unipi unikernels answer requests on both their IPv4 and IPv6 addresses of their tap interfaces in dual-stack OpenBSD hosts,

indeed :)

> and uses IPv6 for DNS resolution

yes :) also using DNS-over-TLS by default (to anycast.uncensoreddns.org)

> as well as HTTPS to fetch the selected branch from Github, if available.

sadly not yet AFAICT (lack of using [happy-eyeballs](https://github.com/roburio/happy-eyeballs) in the git client code)

**Comment by hannesm** [comment]
merged manually into main, thanks for the PR!

**Comment by hannesm** [comment]
It has been some months after merging this, though there is some regression:
- If tls/https was configured, previously a redirect from http to http (port 80 to 443) was in place - this PR removed that one (re-added in f2825c396453fa61912146675fd7c28750f77355
- The redirect function was wrong, previously it used "let uri = Cohttp.Request.uri request", which is a full uri. Now it uses "let path = request.Httpaf.Request.target" -- which is only the path part of the uri, the result is that the redirect of `http://10.0.42.2/foo.html` puts the location to `https:/foo.html` -- i.e. missing host. (fixed in 6ab8f1fdb8dbaf93be23dfdc2365c3c97d77290f)
- The KV lookup used `let p = Uri.path (Cohttp.Request.uri request) in`, which is the path, and only the path (i.e. no query parameters), now `let path = request.Httpaf.Request.target in` is used, which again is the path including query parameters -- so `https://10.0.42.2/foo.html?v=23` leads to a not found instead of delivering `foo.html`. (fixed in 91d0260fa968bfb42ccb8919bdbacfe886632b4f)

**Pull request closed without merging**